docs/local_assembly.tex

```diff
@@ -0,0 +1,121 @@
+
+\begin{figure}
+\center
+\includegraphics[scale=0.5]{zipped.png}
+\caption{A zipped sequence graph.}
+\label{fig:zipped}
+\end{figure}
+
+\item merging diamonds: The assembler looks for nodes $A$ and $C$ such that multiple paths $A \rightarrow B_i \rightarrow C$ exist and absorbs the common prefix of $\{B_i\}$ into $A$ and the common suffix of $\{ B_i \}$ into $C$.  For example, if $k = 10$ and there was a single SNV bubble in the de Bruijn graph, the zipped sequence graph has a reference source path  ($A$), a reference sink path ($C$) and two sides of the bubble ($B_1$ and $B_2$) that are each $10$ bases long.  Since $B_1$ and $B_2$ differ only in a single base, their common bases can be absorbed into $A$ and $C$ such that $B_1$ and $B_2$ contain only a single base each.  Merging tails works the same way without the terminal vertex $C$ and merging common suffixes works the same way without the terminal vertex $A$.  Figure \ref{fig:diamond} shows the graph of Figure \ref{fig:zipped} after the common suffix GGTT of the two sides of the bubble are absorbed into the terminal vertex.
+
+\begin{figure}
+\center
+\includegraphics[scale=0.5]{diamond.png}
+\caption{A merged diamond.}
+\label{fig:diamond}
+\end{figure}
+
+\item splitting common prefixes and suffixes: If there are multiple nodes $S_i$ with a common predecessor and successor of the form $S_i = A + x_i + B$, where $A$ is a common prefix, $B$ is a common suffix, and $x_i$ is unique to $S_i$, the assembler splits each $S_i$ such that $A$ and $B$ become independent nodes, with the $x_i$ in between.
+
+\item The clean-up steps of zipping chains, merging diamonds, tails, and common suffixes, and splitting common prefixes and suffixes are repeated until no more transformations occur.
+
+\end{itemize}
+
+When the \code{debugGraphTransformations} argument is used, the assembly engine emits graphviz .dot files for each of the steps above.
+
+\section{Finding Haplotypes} \label{finding-haplotypes}
+At this point there is one cleaned sequence graph for each successful kmer size.  For each graph we find the best haplotypes\footnote{This is 128 by default and set by the \code{maxNumHaplotypesInPopulation} argument.} according the following score: the score of a path (haplotype) in a sequence graph is the sum over all branching vertices in the path of the log of the multiplicity of the outgoing edge in the path minus the log of the total multiplicity of all outgoing edges.  This formula is implemented as a recursive depth-first search: Starting from the reference source vertex, the haplotype finder adds branching scores and instantiates new haplotype finders for each edge whenever a path diverges and otherwise simply moves to the succeeding vertex without changing the accumulated score.  The procedure ends when every haplotype finder reaches a sink vertex, that is, when the exhaustive search of all paths is finished.
```

Comment:
It uses dynamic programing where the reused subproblems are the haplotype suffixes.... I'm not sure that depth-first is an accurate descriptor... it seems that you are describing the first algorithm that was use (leaving their also different scoring system aside). 

"The procedure ends when every haplotype finder reaches a sink vertex, that is, when the exhaustive search of all paths is finished" That seem measleading... It never explicitly enumerates all possible haplotypes... in some extraordinary cases these could be more than Intger.MAX_VALUE haplotipes. 
The algoritm figures out the best K haplotypes buy reusing subproblem computation. 

The current algorithm has a few short comings... First the code is more complex that it needs to be ... then there is an alternative algorithm base of dijstra that can produce the best K-paths just as efficiently. 

Also is has been shown that taking the K best haplotypes won't necessarely recover all plaussible variation; so this could also be improved.

Reply:
> I'm not sure that depth-first is an accurate descriptor

It seems to me that *if not for caching* the search would be depth-first, that is, the recursion essentially has the form
```
find(vertex) {
   for (edge in outgoing edges) {
      find(target(edge));
   }
}
```
What I mean is that I don't want to just say that it's dynamic programming because that explains the caching but not the traversal.  This matters because if you implemented it, say, as breadth-first search you would get an explosion on sub-problems with no benefit from the caching.

I certainly agree that it doesn't explicitly enumerate the haplotypes.  How about I say that it obtains equivalent results to an exhaustive depth-first search but greatly reduces the cost by caching the best scores starting from every previously considered vertex?